File: AutomataConverter/src/main/kotlin/NonFiniteAutomata.kt
package AutomataConverter

import java.util.*

class NonFiniteAutomata(cardinality: Int,
                        validTokens: String,
                        acceptingStates: Array<Int>,
                        startState: Int,
                        transitionMap: List<Transition>
) : Automata(cardinality, validTokens, acceptingStates, startState, transitionMap) {
    companion object Parser
    {
        fun parse(text: String) : NonFiniteAutomata
        {
            val lines = text.trim().split("\n")

            var i = 0
            val cardinality = lines[i++].toInt()
            val validTokens = lines[i++]
            val acceptingCardinality = lines[i++].toInt()
            val acceptingStates = Array(acceptingCardinality, { 0 })

            var idx = 0
            while(idx < acceptingCardinality)
            {
                acceptingStates[idx++] = lines[i++].toInt()
            }

            val startState = lines[i++].toInt()
            val transitionMap = ArrayList<Transition>()
            while(i < lines.size)
            {
                val transition = lines[i++].split(" ")
                if(transition[1].length > 1) throw IllegalStateException("Transition state must be composed of a single character (got ${transition[1]})")

                val from = transition[0].toInt()
                val to = transition[2].toInt()

                transitionMap.add(
                        Transition(
                                State(from, isStartState = from == startState, isAcceptingState = acceptingStates.contains(from)),
                                transition[1][0],
                                State(to, isStartState = to == startState, isAcceptingState = acceptingStates.contains(to))
                        )
                )
            }

            return NonFiniteAutomata(cardinality, validTokens, acceptingStates, startState, transitionMap)
        }
    }
}

File: src/AutomataConverter/NonFiniteAutomata.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace AutomataConverter
{
    public class NonFiniteAutomata : Automata
    {
        public NonFiniteAutomata(int cardinality, string tokens, IEnumerable<int> acceptingStates, int startState, Dictionary<int, IEnumerable<Transition>> transitionMap)
            : base(cardinality, tokens, acceptingStates, startState, transitionMap)
        {
        }

        public static NonFiniteAutomata parse(string source)
        {
            var lines = new Regex("\\s+").Split(source.Trim());

            var i = 0;
            var cardinality = int.Parse(lines[i++]);
            var validTokens = lines[i++];
            var acceptingCardinality = int.Parse(lines[i++]);
            var acceptingStates = new List<int>(acceptingCardinality);

            for(var idx = 0; idx < acceptingCardinality; idx++)
            {
                acceptingStates.Add(int.Parse(lines[i++]));
            }

            var startState = int.Parse(lines[i++]);
            var transitionMap = new List<Transition>();
            while(i < lines.Length)
            {
                var from = int.Parse(lines[i++]);
                var via = lines[i++];

                if(via.Length > 1) throw new ArgumentOutOfRangeException("via", "Transition string cannot be more than one character");

                var to = int.Parse(lines[i++]);

                transitionMap.Add(new Transition(from, via[0], to));
            }

            return new NonFiniteAutomata(cardinality, validTokens, acceptingStates, startState, transitionMap.GroupBy(t => t.From).ToDictionary(k => k.Key, v => v.AsEnumerable()));
        }
    }
}
